File: EBNFParser/ObjectRegex/Node.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Sep 11 17:49:50 2017

@author: misakawa
"""

import re
DEBUG = False
class MetaInfo(dict):
    
    def __init__(self, count=0 , rdx=0, trace = None):
        self['count'] = count
        self['trace'] = trace if trace else []
        self['rdx']   = rdx
        self['last_count'] = 0
        self['history'] = []
    
    
    @property
    def branch(self):
        self['history'].append((self.count, self.rdx, len(self.trace)))
        return self
    
    @property
    def rollback(self):
        try:
            count, rdx, length = self['history'].pop()
        except IndexError:
            return None
        self.count = count
        self.rdx   = rdx
        self.trace = self.trace[:length]
        return self
    
    @property
    def pull(self):
        try:
            self['history'].pop()
        except IndexError:
            raise Exception("pull no thing")
        return self
    
        
    
    @property
    def last_count(self):
        return self['last_count']
    
    @last_count.setter
    def last_count(self, v):
        self['last_count'] += v
        return self
    

  
    
    @property
    def count(self):
        return self['count']
    
    @count.setter
    def count(self,v):
        self['count'] = v
        return self
    
    @property
    def trace(self):
        return self['trace']
    
    @trace.setter
    def trace(self,v):
        self['trace'] = v
        return self
    
    @property
    def rdx(self):
        return self['rdx']
    
    @rdx.setter
    def rdx(self,v):
        self['rdx'] = v
        return self
    
def reMatch(x, make = lambda x:x, escape = False):
    
    re_ = re.compile( re.escape(x) if escape else x )
    def _1(y):
        if not y: return None
        r = re_.match(y)
        if not r : return None
        a, b = r.span()
        if a is not 0 : raise Exception('a is not 0')
        if b != len(y):
            return None
        return y
    return _1

class Liter:
    def __init__(self, i, name = None):
        self.f = reMatch(i)
        self.name = name
        self.has_recur = False
    def match(self, objs, meta_info = None, partial = True):
        if not meta_info: meta_info = MetaInfo()
        if not objs[meta_info.count:]: return None
        r = self.f(objs[meta_info.count])
        if r is None or (not partial and len(objs) != 1):
            return None
        
        if r == '\n':
            meta_info.rdx += 1
        meta_info.count += 1
        
        return r
    
class ELiter:
    def __init__(self, i, name = None):
        self.f = reMatch(i, escape = True)
        self.name = name
        self.has_recur = False
    def match(self, objs, meta_info = None, partial = True):
        if not meta_info: meta_info = MetaInfo()
        if not objs[meta_info.count:]: return None
        r = self.f(objs[meta_info.count])
        if r is None or (not partial and len(objs) != 1):
            return None
        
        if r == '\n':
            meta_info.rdx += 1
        meta_info.count += 1
        
        return r
    
class recur:
    def __init__(self, name):
        self.name = name


def redef(self, *args, **kwargs):
    self.__init__(*args, **kwargs)
    return self
    
class mode(list):
    def setName(self, name):
        self.name = name
        return self
    
def handle_error(func):
    def _f(objs, meta_info = None, partial = True):
        if not meta_info: 
            meta_info = MetaInfo()
        if  meta_info.trace:
            return func(objs, meta_info = meta_info, partial = partial)
        res = func(objs, meta_info = meta_info, partial = partial)
        if res is None:
            c = meta_info.count
            r = meta_info.rdx
            for ch in objs[c:]:
                if ch is '\n':
                    r += 1
                    c += 1
                    continue
                break
            info = " ".join(objs[c:c+10])
            if len(objs)>c+10:
                info += '...'
            raise SyntaxError(f'Syntax Error at row {r} <= {info}')
        return res
    return _f

class ast:
    def __init__(self, compile_closure, *ebnf, name = None):
        self.name     = name
        self.possibilities= []
        self.has_recur = False
        self.cache    = ebnf
        self.compile_closure = compile_closure
        self.compiled = False
        
    @property
    def compile(self):
        if self.compiled: return self
        for es in self.cache:
            self.possibilities.append([])
            for e in es:
                if e is recur:
                    self.possibilities[-1].append(self)
                    if not self.has_recur:
                        self.has_recur = True
                elif isinstance(e, recur):
                    e = self.compile_closure[e.name]
                    self.possibilities[-1].append(e)
                elif isinstance(e, ast):
                    e.compile
                    self.possibilities[-1].append(e)
                else:
                    self.possibilities[-1].append(e)
        del self.cache
        self.compiled = True
        return self
    
    def match(self, objs, meta_info = None, partial = True):
        if not meta_info: 
            meta_info = MetaInfo()
        if DEBUG:
            print(f'{self.name} WITH {meta_info.trace}')
        res   = mode().setName(self.name)
        for possible in self.possibilities:
            meta_info.branch
            for thing in possible:
                history_i = (meta_info.count, thing.name)
                if thing.has_recur:
                    
                    if history_i in meta_info.trace:
                        print("Found L-R! Dealed!")
                        r = None
                    else:
                        meta_info.trace.append(history_i)
                        r = thing.match(objs, meta_info=meta_info)
                else:
                    meta_info.trace.append(history_i)
                    r = thing.match(objs, meta_info=meta_info)
                
                if r is None:
                    # nexr possible
                    res.clear()
                    meta_info.rollback
                    break
                
                if isinstance(thing, Seq):
                    res.extend(r)
                else:
                    res.append(r)
                if DEBUG:
                    print(f"{thing.name} <= {r}")
            else:
                break
            continue
        else:
            if DEBUG:
                print('RET\n')
            return None
        meta_info.pull
        if DEBUG:
            print('RET\n')
        if partial or meta_info.count == len(objs):
            if DEBUG:
                print('RET\n')
            return res
        if DEBUG:
            print('RET\n')
        return None

    
class Seq(ast):
    def __init__(self,compile_closure, *ebnf, name = None, atleast = 1):
        super(Seq, self).__init__(compile_closure, *ebnf, name = name)
        self.atleast = atleast
    def match(self, objs, meta_info=None, partial = True):
        if not meta_info: 
            meta_info = MetaInfo()
        if DEBUG:
            print(f'{self.name} WITH {meta_info.trace}')
        res   = mode().setName(self.name)
        if not objs[meta_info.count:]:
            if self.atleast is 0:
                return res
            return None
        meta_info.branch
        while True:
            r = super(Seq, self).match(objs, meta_info = meta_info)
            if r is None:
                break
            res.extend(r)
        if DEBUG:
            print(f"{self.name} <= {r}")
        if len(res) < self.atleast:
            meta_info.rollback
            return  None
        meta_info.pull
        return res
            
            
            
            
            
            
            
            

File: CSharp/ObjectRegex/Node.cs
#undef DEBUG
using System.Text.RegularExpressions;
using System;
using System.Runtime.Serialization;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace CSharp.ObjectRegex
{

    public class ObjRegexError : Exception
    {
        public ObjRegexError(string info) : base(info)
        {
        }
    }
    public class StackOverFlowError: Exception{
        public StackOverFlowError(string info) : base(info)
        {
        }
    }
    public class NameError: Exception{
        public NameError(string info) : base(info)
        {
        }
    }
    public class CompileError: Exception{
        public CompileError(string info) : base(info)
        {
        }
    }
    public class RegexTool{
         
        public static (string, Func<string, string>) reMatch(string a, bool escape = false){
            var token_rule = escape?Regex.Escape(a):a;    
            var re_        = new Regex( token_rule );
    
            return (token_rule, 

                   (string str)=>{
                        if (str.Length.Equals(0))
                            return null;
                        var r = re_.Match(str);
                        if (r.Index != 0 || r.Length != str.Length){
                            return null;
                        }
                        return str;
                        }
                    );
        } 
    } 
    public class Mode :List<Mode>{
        
        
        public string name = null;
        public string value= null;
    
        public Mode setName(string name){
            this.name = name;
            return this;
        } 
        public Mode setValue(string value){
            this.value = value;
            return this;
        }
  
        public string Dump(int i=1){
            var space = "\n"+new String(' ', 4*i);
            var toDump = string.Join("", 
                this.Select(
                    mode =>{
                        if (mode.value != null){
                            if (mode.value.Equals("\n")){
                                return $"{space}{mode.name}['{@"\n"}']{space}";
                            }
                            return $"{mode.name}['{mode.value}']{space}";
                        }
                        else{
                            
                            return mode.Dump(i+1);
                        }
                            
                    }
            ));
            return $"{this.name}[{toDump}{space}]";
        }
    }

    public class MetaInfo{
        public int count = 0;
        public int rdx = 0;

        public int trace_length = 0;
        public int history_length = 0;
        public Stack<(int, string)> trace  = null;
            

        public Stack<(int,int,int)> history = null;

        protected (int, string) tracePop(){
            --trace_length;
            return trace.Pop();
        }

        public void tracePush((int, string) tp){
            ++trace_length;
            trace.Push(tp);
        }

        protected (int, int, int) historyPop(){
            --history_length;
            return history.Pop();
        }

        protected void historyPush((int, int, int) tp){
            ++history_length;
            history.Push(tp);
        }

        public MetaInfo(int count = 0, int rdx = 0, Stack<(int, string)> trace = null){
            this.count   = count;
            this.rdx     = rdx;

            if (trace == null )
            {
                this.trace = new Stack<(int, string)>();
            }
            else{
                this.trace = trace;
                foreach(var i in trace)
                {
                    ++this.trace_length;            
                }
            }

            this.history = new Stack<(int,int,int)>();
        }
        public MetaInfo branch (){           
            historyPush((count, rdx, trace_length)); 
            return this;
        }
        public MetaInfo rollback(){
            if (history_length == 0){
                throw new StackOverFlowError("Pull nothing.");
            }
            int trace_length;
            (this.count, this.rdx, trace_length) = historyPop();
            while (trace_length != this.trace_length){
                tracePop();
            }
            return this;
        }
        public MetaInfo pull(){

            if (history_length == 0){
                throw new StackOverFlowError("Pull nothing.");
            }

            historyPop();            
            return this;
        }
        public string DumpTrace(){
            return String.Join(" ", 
                trace.Select((a, b)=> $"({a}, {b})")
            );
        }
    };

    public abstract class BaseAst{
        public string name;
        public bool   has_recur;

        public abstract Mode Match(string[] objs, ref MetaInfo meta, bool partial = true);

    }

    public class LazyDef :BaseAst{

        public LazyDef(string name){
            this.name = name;
        }
        public override Mode Match(string[] objs, ref MetaInfo meta, bool partial = true){
                throw new  CompileError($"Ast {this.name} Not compiled!");
        }
    
    }
    public class Liter : BaseAst{
        public string token_rule;

        public Func<string, string> f;


        public Liter(string i, string name = null, bool escape = false ){

            var (token_rule, f) = RegexTool.reMatch(i, escape);
            this.name = name;
            this.has_recur = false;
            this.token_rule = token_rule;
            this.f = f;
        }
        public override Mode Match(string[] objs, ref MetaInfo meta, bool partial = true){

            int left = objs.Length - meta.count ;
            if (left==0) return null;
            var r = f(objs[meta.count]);  
            if ( r == null || ( (!partial)&&(left !=1 ))){
                    return null;
            }
            if (r == "\n")
                meta.rdx += 1;
            meta.count += 1;
            
            return new Mode().setName(this.name).setValue(r);
        }
        public static Liter ELiter(string i, string name = null){
            
            return new Liter(i, name, true);
        }
    }
    public class Ast : BaseAst{
        public List<List<BaseAst>> possibilities; 
        protected BaseAst[][] cache;
        public Dictionary<string, Ast> compile_closure;
        public bool compiled ;

        public Ast(){

        }
        public Ast(ref Dictionary<string, Ast> compile_closure,
                   string name = null,
                   params BaseAst[][] ebnf){
            
            if (name == null)
                throw new NameError("Name not found! Each kind of ast should have a identity name.");
            this.name = name;

            cache = ebnf;
            if (compile_closure.Keys.Contains(name)){
                throw new NameError("Name of Ast should be identified!");
            }
            compile_closure[name] = this;
            compiled = false;
            this.compile_closure = compile_closure;
            possibilities = new List<List<BaseAst>>();
        }

        public Ast compile(ref HashSet<string> recur_searcher){
            
            if (recur_searcher == null){
                recur_searcher = new HashSet<string>();
                recur_searcher.Add(name);
                
            }
            else{
                if (recur_searcher.Contains(name)){
                    has_recur = true;
                    compiled  = true;
                }
                else{
                    recur_searcher.Add(name);
                }
            }

            if (compiled) {  
                return this; 
            }
            
            foreach(var es in cache){
                var possibility = new List<BaseAst>();
                possibilities.Add(possibility);
                foreach(var e in es){
                    if (e is Liter){
                        possibility.Add(e);
                    }
                    else{
                        
                        if (e is LazyDef){
                            Ast refered = compile_closure[e.name];
                            refered.compile(ref recur_searcher);
                            possibility.Add(refered);
                            if (refered.has_recur)
                                has_recur = true;
                        }
                        else if (e is Seq){
                            Seq refered = e as Seq;
                            refered.compile(ref recur_searcher);
                            possibility.Add(refered);
                            if (refered.has_recur)
                                has_recur = true;
                        }
                        else{
                            throw new CompileError("Unsolved Ast Type");
                        }
                        
                    }
                }
                    
            }
            #if DEBUG
                if (has_recur)
                    Console.WriteLine("Found Recursive EBNF Node => "+name);
            #endif
            
            cache = null;
            compiled = true;
            return this;
        }

        public override Mode Match(string[] objs, ref MetaInfo meta, bool partial = true){
            if (meta == null){
                throw new ArgumentNullException(nameof(meta));
            }
            #if DEBUG
                Console.WriteLine($"{this.name} ");
            #endif

            var res     = new Mode().setName(name);
            var has_res = false; 

            foreach(var possibility in possibilities){
                meta.branch(); // Make a branch in case of resetting.
                foreach(var thing in possibility){
                    var history = (meta.count, thing.name);
                    Mode r;

                    
                    if (thing.has_recur){
                        if (meta.trace.Contains(history)){
                            Console.WriteLine("Found Left Recursion. Dealed.");
                            r = null;
                        }
                        else{
                            meta.trace.Push(history);
                            r = thing.Match(objs, ref meta, true);
                        }
                    }
                    else{
                        // DEBUG View: the trace of parsing.
                        #if DEBUG
                            meta.trace.Push(history);
                        #endif
                        r = thing.Match(objs, ref meta, true);
                    }
                    
                    if (r == null){ 
                        // Do not match current possibility, then rollback.
                        res.Clear();
                        meta.rollback();
                        goto ContinueForNewPossibility;
                    }

                    if (thing is Seq){
                        res.AddRange(r);
                    }
                    else{
                        res.Add(r);
                    }

                    #if DEBUG
                        Console.WriteLine($"{thing.name} <= {r.Dump()}");
                    #endif
                }
                #if DEBUG
                    Console.WriteLine($"RETURN from {this.name} ");
                #endif
                has_res = true;
                break;
                ContinueForNewPossibility:;
            }
            if (!has_res){
                #if DEBUG
                    Console.WriteLine($"RETURN None from {this.name} ");
                #endif
                return null;
            }
            meta.pull();
            var left = objs.Length -meta.count;
            if (partial || left == 0){
                #if DEBUG
                    Console.WriteLine($"RETURN Some from {this.name}");
                #endif
                return res;
            }
            #if DEBUG
                Console.WriteLine($"RETURN None from {this.name} (No partial and do not match all)");
            #endif
            return null;
        }
    }
    public class Seq : Ast{
        int atleast;
        int atmost ;
        public Seq(){
            
        }
        public Seq(ref Dictionary<string, Ast> compile_closure,
                   string name = null,
                   int atleast =   1,
                   int atmost  =  -1,
                   params BaseAst[][] ebnf):base(ref compile_closure, name, ebnf){
        this.atleast = atleast;
        this.atmost  = atmost;
        }

        public override Mode Match(string[] objs, ref MetaInfo meta, bool partial = true){
            if (meta == null){
                throw new ArgumentNullException(nameof(meta));
            }
            var res  = new Mode().setName(this.name);
            int left = objs.Length - meta.count;
            if (left == 0){
                return atleast==0?res : null;
            } 
            meta.branch();
            Mode r;
            int idx = 0;
            if (atmost>0){
                while (true){
                    if (idx>=atmost) break;
                    r = base.Match(objs, ref meta, true);
                    if (r == null)
                        break;
                    res.AddRange(r);
                    ++idx;
                }
            }
            else{
                while (true){
                    r = base.Match(objs, ref meta, true);
                    if (r == null) break;
                    res.AddRange(r);
                    ++idx;
                }
            }
            #if DEBUG
                Console.WriteLine($"{name} <= {res.Dump()}");
            #endif
            if (idx<atleast){
                meta.rollback();
                return null;
            }
            meta.pull();
            return res; 
        }
    }
}